### ISSUE TITLE
ERROR: Error setting hidden attribute for ~TMPCLP216091

### ISSUE THREAD

**Issue by HughWarrington** [joyfullservice/msaccess-vcs-addin#535]
I'm using VCS 4.0.34 in Access 365.

I've exported my database (Access 2000 format .mdb) successfully, then I click Build As... to build it into a new separate file. I get an error in Build.log as follows:

```
-------------------------------------
Importing hidden attributes...
  hidden-attributes.json

-------------------------------------
ERROR: Error setting hidden attribute for ~TMPCLP216091
Error 2950: Reserved Error Source: clsDbHiddenAttribute
-------------------------------------
[1] hidden attributes processed.
-------------------------------------
```

hidden-attributes.json looks like this:

```json
{
  "Info": {
    "Class": "clsDbHiddenAttribute",
    "Description": "Database objects hidden attribute"
  },
  "Items": {
    "Scripts": [
      "~TMPCLP216091"
    ]
  }
}
```

I can't work out what this `~TMPCLP216091` thing is. It doesn't appear in the Access Navigation pane, and I can't find any documentation on the internet about what 'Scripts' means in this context.

Any suggestions?

**Comment by bclothier** [comment]
It implies that somehow you committed a copy that was meant to be deleted. Whenever you delete an object from Access application, it doesn't actually delete the blob right away but rename it as `~TMPXXXX` which is then deleted on the compact. Somehow that got exported when it shouldn't have been. I would remove that from the source.

**Comment by HughWarrington** [comment]
I've found part of the answer. 'Scripts' means macros. Not sure why macros are referred to this way, perhaps for historic reasons. Relevant code is [here](https://github.com/joyfullservice/msaccess-vcs-addin/blob/8c9f86d22ff8503e77c1e6163886bf6bb64ea5e9/Version%20Control.accda.src/modules/clsDbHiddenAttribute.cls#L291).

In my database I've looked in the navigation pane after setting the option to view hidden objects, and I can't see ~TMPCLP216091. That string doesn't appear anywhere else in the plugin's export of my database, just in the hidden-attributes.json.

**Comment by HughWarrington** [comment]
@bclothier thanks for the reply, I just missed seeing it before I posted my own.

That's useful to know. If the `~TMP` macro had been exported, wouldn't I expect to find a corresponding file in the macros subfolder of the export? I don't see that.

I tried a 'compact and repair' immediately before exporting again, and I still get the entry in hidden-attributes.json and no corresponding file in the macros folder. It doesn't get mentioned in the Export.log either.

**Comment by bclothier** [comment]
I am not exactly sure about the sequence but the fact you have an entry in the hidden attributes indicates it was at one point exported, then the object was then deleted as part of the compact & repair operation. Ergo, there's nothing to export. 

Therefore, the fix in this case is to delete the JSON file from the git repository. We would need to investigate more to find out how to reproduce this state where a JSON file was generated while the original object was deleted.

**Comment by bclothier** [comment]
Just as FYI - as part of the deletion, the hidden attribute is set on that object. That's why it would get exported but in this case, it shouldn't have been since it's meant to be deleted. 

[Some more information about this behaivor](https://isladogs.co.uk/recover-deleted-objects/).

**Comment by HughWarrington** [comment]
Seems like actually `~TMP` objects can accumulate inside an Access database and get into an undeletable state, even after compact & repair. There's a great post from Colin Riddington about it [here](https://www.access-programmers.co.uk/forums/threads/how-to-remove-deleted-database-objects-from-the-msysobjects-table.293579/).

Following his steps worked for me. I used the less scary method of creating a new macro with the exact same name `~TMPCLP216091`, then saving it and running from the immediate window:

```
DoCmd.DeleteObject acMacro, "~TMPCLP216091"
```

I then ran another compact & repair for good measure. After a new export, the rogue entry no longer appears in hidden-attributes.json, and a subsequent import doesn't generate the error.

Perhaps the plugin could warn/error/skip if it finds it is being asked to export objects named `~TMPxxxx`, but since this is relying on undocumented internal Access behaviour I don't know how we could ever have the confidence to do so. Hopefully in future people will at least find this thread if they need to.

I'll close this as another 'Access nasty that you just need to know about'.
